### PR TITLE
perlPackages.LaTeXML: patch shebangs to use full path to perl on darwin

### LINKED ISSUE #105249: latexml: binary fails to find LaTeXML::Core
The command `latexml -h` (or any other invocation of `latexml`) fails, with

    Can't locate LaTeXML/Core.pm in @INC (you may need to install the LaTeXML::Core module) (@INC contains: /nix/store/8v612layw8hrwvl7wrqw5p0va8fs5wkv-perl5.32.0-LaTeXML-0.8.4/bin/../lib /nix/store/1dh52zfd3ifglz289i827afnxyipgb6s-perl-5.32.0/lib/perl5/site_perl /nix/store/4l3jrimlvzqzvj0964124h5g6h0xic70-perl5.32.0-Archive-Zip-1.68/lib/perl5/site_perl /nix/store/1pj1sb8diybb5975gxyzni1kznnnpp8y-perl5.32.0-DB_File-1.853/lib/perl5/site_perl /nix/store/2gn46b896k9ca78pqxz28hbcz80plzwg-perl5.32.0-File-Which-1.23/lib/perl5/site_perl /nix/store/zn1xy74fzr3m87zm5an8dpb6cayk7vxb-perl5.32.0-IO-String-1.08/lib/perl5/site_perl /nix/store/x9wa34pcwrpa66m63zbbiqpgv4nldky2-perl5.32.0-Image-Size-3.300/lib/perl5/site_perl /nix/store/2ln30mjzlyd8hv84j2ajaz4cvhs7c909-perl5.32.0-JSON-XS-4.02/lib/perl5/site_perl /nix/store/5sy7n9n9prr5dkwy6dd42z93m4d9vs9z-perl5.32.0-Types-Serialiser-1.0/lib/perl5/site_perl /nix/store/pvc5c61vbfb1aaxcrxmysi0m37kixx5m-perl5.32.0-common-sense-3.75/lib/perl5/site_perl /nix/store/2cjv2r047s2qqz9pvnsg368v44lkjyzq-perl5.32.0-libwww-perl-6.47/lib/perl5/site_perl /nix/store/w48gyrd7v0fl15b3mgbsba2vszyn5cma-perl5.32.0-File-Listing-6.04/lib/perl5/site_perl /nix/store/a36sxw8w2npbyvvp4dgd6xwmkzpv28nl-perl5.32.0-HTTP-Date-6.05/lib/perl5/site_perl /nix/store/jkg855zrj2r448iva7figbvlxm2x62l9-perl5.32.0-TimeDate-2.33/lib/perl5/site_perl /nix/store/g6hagv1ba3d94m3qfldkvhd1pw8sv056-perl5.32.0-HTML-Parser-3.75/lib/perl5/site_perl /nix/store/arfaihzawl93nr143b4i2gzrg3p1vikj-perl5.32.0-HTML-Tagset-3.20/lib/perl5/site_perl /nix/store/3f6ad5k3x9fvyky7g1g2h64s7llvdij0-perl5.32.0-HTTP-Message-6.25/lib/perl5/site_perl /nix/store/y694z6l0i3xn1c59znf0jg8pmjrcxq14-perl5.32.0-Encode-Locale-1.05/lib/perl5/site_perl /nix/store/92188lhs8brqca5vyvns829n4kvns9kb-perl5.32.0-IO-HTML-1.001/lib/perl5/site_perl /nix/store/kzj59zs746wwdwwdhgfpan78lsx5b8w3-perl5.32.0-LWP-MediaTypes-6.04/lib/perl5/site_perl /nix/store/s5wp8n5dkcxb9zkw4pjgw346x4gr02ka-perl5.32.0-URI-1.76/lib/perl5/site_perl /nix/store/b52c76jkxf39cm6rr1rf18mklnz246yh-perl5.32.0-HTTP-Cookies-6.08/lib/perl5/site_perl /nix/store/00lailc90xh1gh26m8a495lks57mrhsz-perl5.32.0-HTTP-Daemon-6.01/lib/perl5/site_perl /nix/store/jrc3kbjpq14q4cvc9dm3xzmdqp47kqvy-perl5.32.0-HTTP-Negotiate-6.01/lib/perl5/site_perl /nix/store/q04iqw4acphc0z8i7xnk2kx3j6gs7j2z-perl5.32.0-Net-HTTP-6.19/lib/perl5/site_perl /nix/store/bz8f1w1xmiz061ihjylfzxcfhmsv1gl9-perl5.32.0-Try-Tiny-0.30/lib/perl5/site_perl /nix/store/a0dybgw5zm9xs4bzrwac8ls6pw8gl6mb-perl5.32.0-WWW-RobotRules-6.02/lib/perl5/site_perl /nix/store/3bcczfrnl4642bwxqmx12ih6pl8mmd89-perl5.32.0-Parse-RecDescent-1.967015/lib/perl5/site_perl /nix/store/3fj4k50sqx8d71wszxiq6nk75kjdkkfd-perl5.32.0-Pod-Parser-1.63/lib/perl5/site_perl /nix/store/mychnpsacrkmfda3prl5jrh4a7yhzk56-perl5.32.0-Text-Unidecode-1.30/lib/perl5/site_perl /nix/store/q93lm9l5csw93hsk2rkzs1v9jnffsww9-perl5.32.0-XML-LibXSLT-1.99/lib/perl5/site_perl /nix/store/zgbxc7lc7gqmpmidxbr3yspa7dgfwzw5-perl5.32.0-XML-LibXML-2.0205/lib/perl5/site_perl /nix/store/xg4f8qwpnh42550x923jybiin4zb134d-perl5.32.0-XML-SAX-1.02/lib/perl5/site_perl /nix/store/fbb4wrmrkqi9wcqly7f6s9hy7z0zcc0v-perl5.32.0-XML-NamespaceSupport-1.12/lib/perl5/site_perl /nix/store/vjbv5032bpjz2jf72rn3f8hs63z7j8wx-perl5.32.0-XML-SAX-Base-1.09/lib/perl5/site_perl /nix/store/8v612layw8hrwvl7wrqw5p0va8fs5wkv-perl5.32.0-LaTeXML-0.8.4/lib/perl5/site_perl /Library/Perl/5.18/darwin-thread-multi-2level /Library/Perl/5.18 /Network/Library/Perl/5.18/darwin-thread-multi-2level /Network/Library/Perl/5.18 /Library/Perl/Updates/5.18.4 /System/Library/Perl/5.18/darwin-thread-multi-2level /System/Library/Perl/5.18 /System/Library/Perl/Extras/5.18/darwin-thread-multi-2level /System/Library/Perl/Extras/5.18 .) at /nix/store/8v612layw8hrwvl7wrqw5p0va8fs5wkv-perl5.32.0-LaTeXML-0.8.4/bin/latexml line 20.
    BEGIN failed--compilation aborted at /nix/store/8v612layw8hrwvl7wrqw5p0va8fs5wkv-perl5.32.0-LaTeXML-0.8.4/bin/latexml line 20.

**To Reproduce**

  1. `nix-shell -p perl532Packages.LaTeXML`
  2. `latexml -h`

**Expected behavior**

I expect to see the `latexml` help text.

**Additional context**

No other context beyond the metadata below.

**Notify maintainers**

Sorry: I'm afraid I can't work out who the maintainers are (there aren't any listed in [the packages entry](https://search.nixos.org/packages?channel=unstable&show=perl532Packages.LaTeXML&from=0&size=30&sort=relevance&query=latexml), and I can't find where the package is in the `nixpkgs` tree at github.

**Metadata**

  - system: `"x86_64-darwin"`
  - host os: `Darwin 19.6.0, macOS 10.15.7`
  - multi-user?: `no`
  - sandbox: `no`
  - version: `nix-env (Nix) 2.3.9`
  - channels(norman): `"nixpkgs-21.03pre254169.4f165c6d7d4"`
  - nixpkgs: `/Users/norman/.nix-defexpr/channels/nixpkgs` 

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by veprbl**
###### Motivation for this change

Fixes: #105249

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
```
ormat error: file is too short
 at /nix/store/swsbd5a4pvzm6sg8dvkxck8qi8vzaasy-perl5.30.3-Archive-Zip-1.68/lib/perl5/site_perl/5.30.3/Archive/Zip/Archive.pm line 1031.
    Archive::Zip::Archive::_findEndOfCentralDirectory(Archive::Zip::Archive=HASH(0x1098168), IO::File=GLOB(0x452c48)) called at /nix/store/swsbd5a4pvzm6sg8dvkxck8qi8vzaasy-perl5.30.3-Archive-Zip-1.68/lib/perl5/site_perl/5.30.3/Archive/Zip/Archive.pm line 761
    Archive::Zip::Archive::readFromFileHandle(Archive::Zip::Archive=HASH(0x1098168), IO::File=GLOB(0x452c48), "931_testor95.epub") called at /nix/store/swsbd5a4pvzm6sg8dvkxck8qi8vzaasy-perl5.30.3-Archive-Zip-1.68/lib/perl5/site_perl/5.30.3/Archive/Zip/Archive.pm line 729
    Archive::Zip::Archive::read(Archive::Zip::Archive=HASH(0x1098168), "931_testor95.epub") called at t/931_epub.t line 35

#   Failed test 'epub file successfully loads as Archive::Zip object'
#   at t/931_epub.t line 35.
#          got: '3'
#     expected: '0'

#   Failed test 'correct number of files were present in final ePub'
#   at t/931_epub.t line 36.
#          got: '0'
#     expected: '10'

#   Failed test 'correct files were present in final ePub: '
#   at t/931_epub.t line 38.

#   Failed test 'log file was written to epub'
#   at t/931_epub.t line 41.
Can't call method "contents" on an undefined value at t/931_epub.t line 42.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 25 just after 7.
```

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 105271` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>perl530Packages.LaTeXML</li>
    <li>perl532Packages.LaTeXML</li>
  </ul>
</details>

**Comment by veprbl** [comment]
@GrahamcOfBorg build perl530Packages.LaTeXML
@GrahamcOfBorg build perl532Packages.LaTeXML

**Comment by veprbl** [comment]
#105283 was merged, so again:
@GrahamcOfBorg build perl530Packages.LaTeXML
@GrahamcOfBorg build perl532Packages.LaTeXML

**Review by stigtsp — APPROVED** [review]
Result of `nixpkgs-review pr 105271` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>perl530Packages.LaTeXML</li>
    <li>perl532Packages.LaTeXML</li>
  </ul>
</details>

Result of `nixpkgs-review pr 105271` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>perl530Packages.LaTeXML</li>
    <li>perl532Packages.LaTeXML</li>
  </ul>
</details>

Tested `bin/latexml*` OK on darwin x64

File: pkgs/top-level/perl-packages.nix [hunk 1]
```diff
@@ -10710,14 +10710,17 @@ let
...
-      shortenPerlShebang $out/bin/latexmlmath
-      shortenPerlShebang $out/bin/latexmlpost
+    propagatedBuildInputs = [ ArchiveZip DBFile FileWhich IOString ImageSize JSONXS LWP ParseRecDescent PodParser TextUnidecode XMLLibXSLT ];
+    preCheck = ''
+      rm t/931_epub.t # epub test fails
+    '';
+    nativeBuildInputs = stdenv.lib.optional stdenv.isDarwin shortenPerlShebang;
+    # shebangs need to be patched before executables are copied to $out
+    preBuild = stdenv.lib.optionalString stdenv.isDarwin ''
+      patchShebangs bin/
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:10720]:
```suggestion
    preBuild = ''
      patchShebangs bin/
    '' + stdenv.lib.optionalString stdenv.isDarwin ''
```

**Review by stigtsp — CHANGES_REQUESTED** [review]
Woops, premature approval.

`bin/latex*` executables fails on linux x64

```
[nix-shell:~/.cache/nixpkgs-review/pr-105271]$ results/perl532Packages.LaTeXML/bin/latexml
/nix/store/93mr9awnj0sb0ca48n1z4ixyyqqrfbbb-coreutils-8.32/bin/env: ‘perl -I/nix/store/03vh3lc5fymqjyck3v3rckbvs36qa74g-perl-5.32.0/lib/perl5/site_perl -I/nix/store/w1kyfhm416v8p8alw5fkd3hxj8d717fv-perl5.32.0-Archive-Zip-1.68/lib/perl5/site_perl -I/nix/st’: No such file or directory
/nix/store/93mr9awnj0sb0ca48n1z4ixyyqqrfbbb-coreutils-8.32/bin/env: use -[v]S to pass options in shebang lines

```

File: pkgs/top-level/perl-packages.nix (hunk 1 above)
Comment (veprbl) [reply pkgs/top-level/perl-packages.nix]: Interesting. Turns out it was broken on Linux too.

**Review by stigtsp — APPROVED** [review]
LGTM

Tested OK on `linux x64` and `darwin x64`

**Pull request merged**